odoo/addons/base/models/res_partner.py

```diff
@@ -449,6 +449,19 @@ def _check_parent_id(self):
         if self._has_cycle():
             raise ValidationError(_('You cannot create recursive Partner hierarchies.'))
 
+    @api.constrains('company_id')
+    def _check_partner_company(self):
+        for partner in self:
+            if partner.is_company and partner.company_id:
+                related_company = self.env['res.company'].search([
```

Comment:
For clarification, can you add a comment: "If there exists an company linked to the partner, the company_id set on the partner must be the company's partner".

Can you rewrite the function so that only one search is done before looping on partners?
Something like:
```python
"""Check that for every partner which has a company, if there exists a company linked to that partner, the company_id set on the partner is that company"""
partners = self.filtered(lambda p: p.is_company and p.company_id)
companies = self.env['res.company'].search_fetch([('partner_id', 'in', partners.ids)], ['partner_id'])
for company in companies:
   if company != company.partner_id.company_id:
     raise ...
```